### DISCUSSION TITLE
Optimizations

### DISCUSSION THREAD

**Discussion by Rocky04** [hathach/tinyusb#2174] [General]
@hathach 

- Why not also compare the [interrupts when fetched with the enabled ones](https://github.com/hathach/tinyusb/blob/8fa0b74d8016554d23f84981f78c656f215d76e8/src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c#L659)? Like `uint32_t int_status = USB->ISTR & USB->CNTR;`

- You [mention here](https://github.com/hathach/tinyusb/blob/8fa0b74d8016554d23f84981f78c656f215d76e8/src/device/usbd.c#L1126), that some devices have issues reacting to a resume signal from the host. I'm aware of such problems on AVR devices and as far as I remember it was host specific. So it should be good if the described failsafe would be used. For example the SOF interrupt can be enabled in the suspend block and disabled again in the resume block. That way the device would not get flooded with SOF interrupts when active while still providing a failsafe to resume normal USB operation when put to sleep.

- I'm somehow skeptical that it's a good idea to skip the rest of the interrupts [when there is a reset](https://github.com/hathach/tinyusb/blob/8fa0b74d8016554d23f84981f78c656f215d76e8/src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c#L679). I would imagine all spurious pending interrupt should be cleared if the return is used. Like what happens when the device had some other things to do and so handle multiple later on. It should be possible that another interrupt was set in addition which then isn't scheduled but a new one can't occur because the old one was never cleared. Or does this somehow not matter?

**Comment by hathach** [comment]
can you make an PR for your suggested change, please make sure to test it on actual hardwware. It is much easier to understand what you mean. I wrote lots of code during long span of time, I don't quite remember everything.
